crates/messages/proto/messages.proto

```diff
@@ -31,137 +19,149 @@ message WorkerState {
 }
 
+// Worker -> Portal
 message QueryResult {
   string query_id = 1;
   oneof result {
-    OkResult ok = 2;
-    string bad_request = 3;
-    string server_error = 4;
-    google.protobuf.Empty no_allocation = 5;
-    google.protobuf.Empty timeout_v1 = 6 [deprecated = true];
-    string timeout = 7;
+    QueryOk ok = 2;
+    QueryError err = 8;
   }
+  optional uint32 retry_after_ms = 9;
 }
 
-message OkResult {
+message QueryOk {
   bytes data = 1;
-  optional bytes exec_plan = 2;
   optional uint64 last_block = 3;
+  QueryResultSummary summary = 4;
 }
 
-message QuerySubmitted {
-  string client_id = 1;
-  string worker_id = 2;
-  string query_id = 3;
-
-  string dataset = 4;
-  string query = 5;
-  bytes query_hash = 6;
+message QueryError {
+  oneof err {
+    string bad_request = 1;
+    string not_found = 2;
+    string server_error = 3;
+    google.protobuf.Empty too_many_requests = 4;
+    google.protobuf.Empty server_overloaded = 5;
+    string timeout = 6;
+  }
 }
 
+// Portal -> Logs collector
 message QueryFinished {
-  string client_id = 1;
   string worker_id = 2;
   string query_id = 3;
 
-  uint32 exec_time_ms = 4;
+  uint32 total_time_ms = 4;
   oneof result {
-    SizeAndHash ok = 5;
-    string bad_request = 6;
-    string server_error = 7;
-    google.protobuf.Empty timeout_v1 = 8  [deprecated = true];
-    google.protobuf.Empty no_allocation = 9;
-    string timeout = 10;
+    QueryResultSummary ok = 5;
+    QueryError err = 11;
   }
 }
 
-message QueryExecuted {
-  string client_id = 1;
-  string worker_id = 2;
-
-  Query query = 3;
-  bytes query_hash = 5;
-
-  optional uint32 exec_time_ms = 6; // optional to enforce serializing 0
-  oneof result {
-    InputAndOutput ok = 7;
-    string bad_request = 8;
-    string server_error = 9;
-  }
-  optional uint64 seq_no = 10; // optional to enforce serializing 0
-  optional uint64 timestamp_ms = 11; // optional to enforce serializing 0
-  bytes signature = 12;
+// Logs collector -> Worker
+message LogsRequest {
+  uint64 from_timestamp_ms = 1;
+  uint64 to_timestamp_ms = 2;
+  optional string last_received_query_id = 3;  // query ID of the last collected query
 }
 
+// Worker -> Logs collector
 message QueryLogs {
   repeated QueryExecuted queries_executed = 1;
+  bool has_more = 2;
 }
 
-message InputAndOutput {
-  optional uint32 num_read_chunks = 1; // optional to enforce serializing 0
-  SizeAndHash output = 2;
-}
+message QueryExecuted {
+  string client_id = 1;
+  Query query = 3;
 
-message SizeAndHash {
-  optional uint32 size = 1; // optional to enforce serializing 0
-  bytes sha3_256 = 2;
+  uint32 exec_time_ms = 6;
+  uint64 timestamp_ms = 11;
+  oneof result {
+    QueryResultSummary ok = 7;
+    QueryError err = 14;
+  }
 }
 
-message LogsCollected {
-  map<string, uint64> sequence_numbers = 1; // peer ID -> last saved seq no
+message QueryResultSummary {
+  uint64 size = 1;
+  bytes hash = 2;  // sha3_256
```

Comment:
If you mean that it can be computed by the client than yes, it can be removed. I thought it's convenient to have this struct reused between the query result and worker logs, but you're right, inlining of the fields would also work fine.